staging/src/k8s.io/pod-security-admission/admission/admission.go

```diff
@@ -0,0 +1,537 @@
+		timeRemaining := time.Duration(0.9 * float64(time.Until(deadline))) // Leave a little time to respond.
+		if timeout > timeRemaining {
+			timeout = timeRemaining
+		}
+	}
+	deadline := time.Now().Add(timeout)
+	ctx, cancel := context.WithDeadline(ctx, deadline)
+	defer cancel()
+
+	pods, err := a.PodLister.ListPods(ctx, namespace)
+	if err != nil {
+		klog.ErrorS(err, "Failed to list pods", "namespace", namespace)
+		return []string{"Failed to list pods"}
+	}
+
+	checks, err := a.Registry.ChecksForLevelAndVersion(enforce.Level, enforce.Version)
+	if err != nil {
+		klog.ErrorS(err, "Failed to get checks for "+enforce.String())
+		return []string{"Policy lookup failed"}
+	}
+
+	var warnings []string
+	if len(pods) > namespaceMaxPodsToCheck {
+		warnings = append(warnings, fmt.Sprintf("Large namespace: only checking the first %d of %d pods", namespaceMaxPodsToCheck, len(pods)))
+		pods = pods[0:namespaceMaxPodsToCheck]
+	}
+
+	for i, pod := range pods {
+		r := policy.AggregateCheckPod(checks, &pod.ObjectMeta, &pod.Spec)
+		if !r.Allowed {
+			// TODO: consider aggregating results (e.g. multiple pods failed for the same reasons)
+			warnings = append(warnings, fmt.Sprintf("%s: %s", pod.Name, r.ForbiddenReason()))
+		}
+		if time.Now().After(deadline) {
+			return append(warnings, fmt.Sprintf("Timeout reached after checking %d pods", i+1))
+		}
+	}
+
+	return warnings
+}
+
+func (a *Admission) PolicyToEvaluate(labels map[string]string) (api.Policy, error) {
+	return api.PolicyToEvaluate(labels, a.defaultPolicy)
+}
+
+// allowResponse is the response used when the admission decision is allow.
+func allowedResponse() admissionv1.AdmissionResponse {
+	return admissionv1.AdmissionResponse{Allowed: true}
+}
+
+// forbiddenResponse is the response used when the admission decision is deny for policy violations.
+func forbiddenResponse(msg string) admissionv1.AdmissionResponse {
+	return admissionv1.AdmissionResponse{
+		Allowed: false,
+		Result: &metav1.Status{
+			Status:  metav1.StatusFailure,
+			Reason:  metav1.StatusReasonForbidden,
+			Message: msg,
+			Code:    http.StatusForbidden,
+		},
+	}
+}
+
+// invalidResponse is the response used for namespace requests when namespace labels are invalid.
+func invalidResponse(msg string) admissionv1.AdmissionResponse {
+	return admissionv1.AdmissionResponse{
+		Allowed: false,
+		Result: &metav1.Status{
+			Status:  metav1.StatusFailure,
+			Reason:  metav1.StatusReasonInvalid,
+			Message: msg,
+			Code:    422,
+		},
+	}
+}
+
+// internalErrorResponse is the response used for unexpected errors
+func internalErrorResponse(msg string) admissionv1.AdmissionResponse {
+	return admissionv1.AdmissionResponse{
+		Allowed: false,
+		Result: &metav1.Status{
+			Status:  metav1.StatusFailure,
+			Reason:  metav1.StatusReasonInternalError,
+			Message: msg,
+			Code:    http.StatusInternalServerError,
+		},
+	}
+}
+
+// isSignificantPodUpdate determines whether a pod update should trigger a policy evaluation.
```

Comment:
Is there a shortcut link to which pod fields are mutable to help with reasoning about whether we got everything?

Reply:
only API validation :-/

https://github.com/kubernetes/kubernetes/blob/release-1.21/pkg/apis/core/validation/validation.go#L3947-L3949